components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt

```diff
@@ -0,0 +1,308 @@
+import com.hotels.styx.api.extension.Origin
+import com.hotels.styx.api.extension.RemoteHost
+import com.hotels.styx.api.extension.service.RewriteRule
+import com.hotels.styx.client.stickysession.StickySessionLoadBalancingStrategy
+import com.hotels.styx.client.stickysession.StickySessionCookie
+import com.hotels.styx.client.retry.RetryNTimes
+import org.reactivestreams.Publisher
+import org.slf4j.LoggerFactory
+import java.lang.StringBuilder
+import java.util.Objects.nonNull
+import java.util.Optional
+
+/**
+ * A configurable HTTP client that uses connection pooling, load balancing, etc.
+ */
+class StyxBackendServiceClient(
+    private val id: Id,
+    rewriteRules: List<RewriteRule>,
+    private val originStatsFactory: OriginStatsFactory,
+    private val originsRestrictionCookieName: String?,
+    private val stickySessionConfig: StickySessionConfig,
+    private val originIdHeader: CharSequence,
+    private val loadBalancer: LoadBalancer,
+    private val retryPolicy: RetryPolicy,
+    private val metrics: CentralisedMetrics,
+    private val overrideHostHeader: Boolean
+) : BackendServiceClient {
+
+    private constructor(builder: Builder): this(
+        id = builder.id,
+        rewriteRules = builder.rewriteRules,
+        originStatsFactory = builder.originStatsFactory,
+        originsRestrictionCookieName = builder.originsRestrictionCookieName,
+        stickySessionConfig = builder.stickySessionConfig,
+        originIdHeader = builder.originIdHeader,
+        loadBalancer = builder.loadBalancer,
+        retryPolicy = builder.retryPolicy,
+        metrics = builder.metrics,
+        overrideHostHeader = builder.overrideHostHeader
+    )
+
+    companion object {
+        private val LOGGER = LoggerFactory.getLogger(StyxBackendServiceClient::class.java)
+        private const val MAX_RETRY_ATTEMPTS = 3
+        inline fun build(id: Id,
+                         originStatsFactory: OriginStatsFactory,
+                         loadBalancer: LoadBalancer,
+                         metrics: CentralisedMetrics,
+                         block: Builder.() -> Unit) = Builder(id,originStatsFactory,loadBalancer, metrics).apply(block).build()
+    }
+
+    /**
+     * A builder for [StyxBackendServiceClient].
+     */
+    class Builder(
+        val id: Id,
+        val originStatsFactory: OriginStatsFactory,
+        val loadBalancer: LoadBalancer,
+        val metrics: CentralisedMetrics
+    ) {
+        var rewriteRules: List<RewriteRule> = emptyList()
+        var originsRestrictionCookieName: String? = null
+        var stickySessionConfig: StickySessionConfig = StickySessionConfig.stickySessionDisabled()
+        var originIdHeader: CharSequence = StyxHeaderConfig.ORIGIN_ID_DEFAULT
+        var retryPolicy: RetryPolicy = RetryNTimes(3)
+        var overrideHostHeader: Boolean = false
+
+        fun build() = StyxBackendServiceClient(this)
+    }
+
+
+    private val rewriteRuleset: RewriteRuleset = RewriteRuleset(rewriteRules)
```

Comment:
This should probably move to the top.

Reply:
do review: https://kotlinlang.org/docs/coding-conventions.html#class-layout, seems like other changes are required. Just makes it easier to read code if everything is organized in the same way. For example, i was wondering why the class doesn't have the 
```
private val rewriteRuleset ...
```
until i reached this line.